### PR TITLE
Fix last_day_of_month for timestamp with time zone

### LINKED ISSUE #2851: Incorrect results for last_day_of_month function
The result is incorrect when a timestamp with time zone  <= UTC 00:00 is provided.
```
presto> select
     ->     last_day_of_month(timestamp '2020-02-17 00:00:00.000 +02:00'),
     ->     last_day_of_month(timestamp '2020-02-17 00:00:00.000 +01:00'),
     ->     last_day_of_month(timestamp '2020-02-17 00:00:00.000 +00:00'),
     ->     last_day_of_month(timestamp '2020-02-17 00:00:00.000 -01:00'),
     ->     last_day_of_month(timestamp '2020-02-17 00:00:00.000 -02:00');
   _col0    |   _col1    |   _col2    |   _col3    |   _col4
------------+------------+------------+------------+------------
 2020-02-29 | 2020-02-29 | 2020-03-01 | 2020-03-01 | 2020-03-01
```

This also affect `last_day_of_month(current_timestamp)`, depends on the timezone of the client.
```
⇒  java  -Duser.timezone=Asia/Tokyo  -jar ./presto-cli-*-executable.jar
presto> select
     ->     current_timestamp,
     ->     last_day_of_month(current_timestamp),
     ->     last_day_of_month(timestamp '2020-02-17 00:00:00.000 Asia/Tokyo'),
     ->     last_day_of_month(timestamp '2020-02-17 00:00:00.000 Asia/Hong_Kong'),
     ->     last_day_of_month(timestamp '2020-02-17 00:00:00.000 Australia/North'),
     ->     last_day_of_month(timestamp '2020-02-17 00:00:00.000 America/New_York'),
     ->     last_day_of_month(timestamp '2020-02-17 00:00:00.000 America/Los_Angeles'),
     ->     last_day_of_month(timestamp '2020-02-17 00:00:00.000 Canada/Atlantic');
               _col0                |   _col1    |   _col2    |   _col3    |   _col4    |   _col5    |   _col6    |   _col7
------------------------------------+------------+------------+------------+------------+------------+------------+------------
 2020-02-17 15:54:29.193 Asia/Tokyo | 2020-02-29 | 2020-02-29 | 2020-02-29 | 2020-02-29 | 2020-03-01 | 2020-03-01 | 2020-03-01

⇒  java  -Duser.timezone=America/Los_Angeles  -jar ./presto-cli-*-executable.jar
presto> select
     ->     current_timestamp,
     ->     last_day_of_month(current_timestamp),
     ->     last_day_of_month(timestamp '2020-02-17 00:00:00.000 Asia/Tokyo'),
     ->     last_day_of_month(timestamp '2020-02-17 00:00:00.000 Asia/Hong_Kong'),
     ->     last_day_of_month(timestamp '2020-02-17 00:00:00.000 Australia/North'),
     ->     last_day_of_month(timestamp '2020-02-17 00:00:00.000 America/New_York'),
     ->     last_day_of_month(timestamp '2020-02-17 00:00:00.000 America/Los_Angeles'),
     ->     last_day_of_month(timestamp '2020-02-17 00:00:00.000 Canada/Atlantic');
                    _col0                    |   _col1    |   _col2    |   _col3    |   _col4    |   _col5    |   _col6    |   _col7
---------------------------------------------+------------+------------+------------+------------+------------+------------+------------
 2020-02-16 22:54:56.653 America/Los_Angeles | 2020-03-01 | 2020-02-29 | 2020-02-29 | 2020-02-29 | 2020-03-01 | 2020-03-01 | 2020-03-01
```

### PR DESCRIPTION
**Pull request by oneonestar**
Fix #2851

### CODE REVIEW

File: presto-main/src/main/java/io/prestosql/operator/scalar/DateTimeFunctions.java [hunk 1]
```diff
@@ -982,7 +982,11 @@ public static long dayFromInterval(@SqlType(StandardTypes.INTERVAL_DAY_TO_SECOND
     public static long lastDayOfMonthFromTimestampWithTimeZone(@SqlType(StandardTypes.TIMESTAMP_WITH_TIME_ZONE) long timestampWithTimeZone)
     {
         long millis = unpackMillisUtc(timestampWithTimeZone);
-        millis = unpackChronology(timestampWithTimeZone).monthOfYear().roundCeiling(millis + 1);
+        ISOChronology isoChronology = unpackChronology(timestampWithTimeZone);
+        millis = isoChronology.monthOfYear().roundCeiling(millis + 1);
+        // millis is currently midnight in timezone of the original value
```
Comment (findepi) [inline presto-main/src/main/java/io/prestosql/operator/scalar/DateTimeFunctions.java]: This comment makes me think about the case when there is no midnight and the date/time is near the end of the month.

For example, `America/Bahia_Banderas` and the local date/time 1970-01-01 00:00:00 ±1ms.

Can you add this to your tests? (along with other zones and date/times you have)

Comment (oneonestar) [reply presto-main/src/main/java/io/prestosql/operator/scalar/DateTimeFunctions.java]: Added tests for timezone `America/Bahia_Banderas` and `2019-01-01 00:00:00.000 ±1ms`.
Not sure how to test for day without midnight. Could you provide a concrete example for this?
`1970-01-01 00:00:00` is epoch and difficult to test.

**Comment by oneonestar** [comment]
The error seems unrelated. How can I trigger a re-test?

**Comment by dain** [comment]
@oneonestar there is a button on the actions page to rerun the jobs, but I'm not sure if you can see it (github permission are weird).  You can always re-push to trigger a rebuild.  I triggered a rebuild.

**Comment by dain** [comment]
@oneonestar Github actions keep failing during the git check out.  I'm not sure if there is something wrong with actions right now, but can you try re pushing your code?

**Comment by oneonestar** [comment]
Still there are some errors that looks like unrelated.

**Comment by dain** [comment]
I tried to rerun those tests, and they fail during check out again.  I can't see the original failures anymore.  Can you re push?  I'll just manually verify next time.

File: presto-main/src/main/java/io/prestosql/operator/scalar/DateTimeFunctions.java [hunk 2]
```diff
@@ -982,7 +982,11 @@ public static long dayFromInterval(@SqlType(StandardTypes.INTERVAL_DAY_TO_SECOND
...
+        // convert to UTC
+        millis = millis + isoChronology.getZone().getOffset(millis) - MILLISECONDS_IN_DAY;
```
Comment (findepi) [inline presto-main/src/main/java/io/prestosql/operator/scalar/DateTimeFunctions.java]: `convertUTCToLocal`

**Comment by findepi** [comment]
Merged as 13e53d785c87eac17dcdda77809330fac4670dd5, thanks!

(i slightly changed comments)

**Pull request closed without merging**